superset/views/core.py

```diff
@@ -1959,7 +1959,7 @@ def dashboard_permalink(  # pylint: disable=no-self-use
         dashboard_id, state = value["dashboardId"], value.get("state", {})
         url = f"/superset/dashboard/{dashboard_id}?permalink_key={key}"
         if url_params := state.get("urlParams"):
-            params = parse.urlencode(url_params)
+            params = parse.urlencode(dict(url_params))
```

Comment:
I'm surprised this wasn't caught by mypy or related tests. Can you double check that the the types are accurate, and see if the tests are correct?

Reply:
Sure

Reply:
@villebro Yes, but at the superset/views/core.py:836 you're parsing url_params the same way using `dict`
```python
        if key is not None:
            command = GetExplorePermalinkCommand(key)
            try:
                permalink_value = command.run()
                if permalink_value:
                    state = permalink_value["state"]
                    initial_form_data = state["formData"]
                    url_params = state.get("urlParams")
                    if url_params:
                        initial_form_data["url_params"] = dict(url_params)
                else:
                    return json_error_response(
                        _("Error: permalink state not found"), status=404
                    )
            except (ChartNotFoundError, ExplorePermalinkGetFailedError) as ex:
                flash(__("Error: %(msg)s", msg=ex.message), "danger")
                return redirect("/chart/list/")
```

Reply:
@villebro, I don't see any test cases that test urlParams in the permalink state.
The type of urlParams is Optional[List[Tuple[str, str]]], which means that parsing urlParams using `dict` is the correct (and works on me)

Reply:
@villebro So the `urlencode` function taking dictionary, it means that we need to make dictionary from urlParams that have type List[Tuple[str, str]] (as you doing it in superset/views/core.py:836)

Reply:
The reason we want to keep it as a list/tuple of tuples is because url params can contain duplicate keys (this would be lost is it were converted to a `dict`). I believe this may be a regression from #23888 , which has lost some context on tuples, which has caused them to become lists, which isn't compatible with `urlencode`.

I think the correct solution is to decode the persisted JSON object with the relevant marshmallow schema, which could be implemented with a new `MarshmallowCodec`. I can open a PR with the required changes. Alternatively, feel free to take a stab at it if you want!

Reply:
Please open PR by yourself!

Reply:
I'll do that, thanks for bringing this to our attention!

Reply:
@Always-prog I was able to reproduce the issue, thanks again!

Reply:
@villebro you are welcome!